### ISSUE TITLE
Bug: The latest update replaces pre-existing configuration files

### ISSUE THREAD

**Issue by sbrl** [ohwgiles/laminar#92]
It's a good thing I take backups, because the armhf .deb file appears to have completely replaced my `/etc/laminar.conf` - regardless of the fact that I've actually changed it since installation.

**Comment by ohwgiles** [comment]
Yes unfortunately I don’t know how to write deb packages. Looking forward to the official Debian packaging

**Comment by sbrl** [comment]
Ah, that would be awesome!

I haven't built .deb packages either, but I think it's a case of marking it (and the systemd service file) as a configuration file. That way it'll prompt if it's different.

**Comment by jakimfett** [comment]
Note that as of https://github.com/ohwgiles/laminar/commit/1642899159103942176fd5aa13139507f16f3204 this bug is **also** affecting build-from-source installations on Debian Linux (v9.9). 

I had modified 5 of the 7 default config options, and the `/etc/laminar.conf` file is overwritten each time `sudo make install` is executed.

**Comment by ohwgiles** [comment]
@jakimfett building from source is a bit of a different picture. Normally, one expects `make install` to overwrite anything in its path, because (traditionally) Makefile rules would just call [`install`](https://linux.die.net/man/1/install). When installing from source, it's good practice to run (as non-root) `make DESTDIR=path/to/tmpdir install`, then carefully curate what should be copied from `tmpdir` into its final location. Or better still, use a package manager; they actually have sensible ways to deal with updated configuration files.

**Comment by ohwgiles** [comment]
Practically speaking, there is a way to avoid this: always install *only* the `laminar.conf.example`. If `laminar.conf` isn't present, laminar should use the same default values that are in the default configuration file. It would be up to the user to copy the example to `laminar.conf`, and modify it as per their needs.
The problem with that idea is that it breaks the feature of package managers to intelligently upgrade the configuration file.

In summary, I don't think that changing the behaviour of `make install` is the right thing to do. However, this issue is valid for the built `.deb` and `.rpm` packages; they should use mark `laminar.conf` as a configuration file, using [this mechanism](https://manpages.debian.org/testing/dpkg-dev/deb-conffiles.5.en.html) and [this mechanism](https://www.cl.cam.ac.uk/~jw35/docs/rpm_config.html) respectively

**Comment by sbrl** [comment]
Have you considered using [fpm](https://github.com/jordansissel/fpm) to build source packages? It has a CLI flag for marking something as a config file.

**Comment by jakimfett** [comment]
> ...When installing from source, it's good practice to run (as non-root) `make DESTDIR=path/to/tmpdir install`, then carefully curate what should be copied...

Functionally, this means that build-from-source becomes a second class citizen in the installation scheme, and requires the end (power)user to do extra steps, introducing human error and variability into the processes.  

I sorta, almost, like the 'practical' option of only ever using `make` to install an example file (which I've seen done in other packages).  
Agreement that borking the user experience (with regard to packages) to 'fix' something in test/dev is a Bad Idea.

Question tho (because despite my sysadmin background, I've never built packages), does the `.rpm` and `.deb` packaging process rely on the `cmake` configuration as well? Eg, one config, three outputs?


I've only just started poking what `cmake` is capable of scripting, and I'm about 60% certain there's a way to have our cake and eat it too.

Moar poking (and attempts to find a decent fix) will happen today as I attempt to make progress on the build servers I'm currently sysadmin-ing at.

---
@sbrl thanks for linking that, I love their stated mentality and approach to "what is a bug".   
I don't think (deferring to the `laminar` project owner obviously) that this particular bug/inconsistency necessitates a rewrite in a different package build paradigm, but depending on what my "moar poking" turns up, I may get annoyed and end up attempting to rewrite more than I originally bargained for.

**Comment by ohwgiles** [comment]
> Functionally, this means that build-from-source becomes a second class citizen in the installation scheme

Yes, sorta. IMO, installing from source is (or at least should be) rare, and requires extra care. I think this is the case for most software.

> Question tho (because despite my sysadmin background, I've never built packages), does the .rpm and .deb packaging process rely on the cmake configuration as well? Eg, one config, three outputs?

Not directly. While there is a mechanism to produce the package from CMake (see [CPack](https://cmake.org/cmake/help/v3.0/module/CPack.html)), I found it inadequate in the past, and for this project wrote a couple of simple shell scripts that build packages in a docker container, see [here](https://github.com/ohwgiles/laminar/blob/bcb8b438f38c9cbd8b6979a031250789685fb507/docker-build-debian.sh) and [here](https://github.com/ohwgiles/laminar/blob/bcb8b438f38c9cbd8b6979a031250789685fb507/docker-build-centos.sh). They just use the results of `make DESTDIR=path/to/tmp install`, so they do indirectly require the CMake `install` directive.

**Comment by ohwgiles** [comment]
Commit 8cf31fa resolves this issue in that subsequent packages built with laminar's internal packaging scripts will trigger apt's configuration file conflict resolution mechanism. This will be part of the next release.
